mautic/mautic
# Fixes #11888 for mailjet puts only : in lead_donotcontact.comment

Fixes #11888



| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | [ x ]
| New feature/enhancement? (use the a.x branch)      | [ ]
| Deprecations?                          | [ ]
| BC breaks? (use the c.x branch)        | [ ]
| Automated tests included?              | [ ] 
| Related user documentation PR URL      | mautic/mautic-documentation#... 
| Related developer documentation PR URL | mautic/developer-documentation#... 
| Issue(s) addressed                     | Fixes #... 

# Description:

## Problem in Mautic instances with mailjet: 

Mautic process the callback response of mailjet for softbounced emails currently so, that in the database field `comment` in the table `lead_donotcontact` only a colon appears / is persisted.

## Reason of this problem and here offered possible solution

Reason of this problem:
So far Mautic uses the mailjet data fields `error_related_to` : `error`.
The JSON mailjet response for these both data fields are currently empty for softbounced emails.
But in the JSON mailjet response are the additional fields "hard_bounce" (boolean) and "comment" (string) included, which are currently ignored by mautic.

The here offered possible solution for this problem:
Checking the JSON response status by evaluate whether event is blocked/bounced, by evaluate whether hard_bounce value is true or false, by including content of relevant mailjet fields,
so that the "comment" field in "lead_donotcontact" will be

`BLOCKED` : `error_related_to` : `error` (if JSON event response status of mailjet is blocked)

`SOFT` : `error_related_to` : `error` : `comment` (if value of hard_bounce in JSON Response of mailjet is false)

and

`HARD` : `error_related_to` : `error` (if value of hard_bounce in JSON-Response of mailjet is true)
   
# Steps to test this PR:

## I.) Reproduce the problem
### To reproduce the problem
1.) In your mautic test instance with mailjet send an email, for example, to a "softbounce simulator" or an email account, which you configured as "full".

2.) Wait until the response of mailjet for the softbounced email is back [Important: This can take up to three days (see also mailjet api documentation), because mailjet retries to send a softbounced email several times until the email will set to the status "softbounced" in mailjet and the callback to mautic will happen]

3.) Open the table `lead_donotcontact` in your mautic test database and you will see, that there is only a colon in the comment field of the softbounced email.

## II.) Testing the solution

### To test the here offered solution 
1.) Implement the patch (with this code changes) in a mautic test instance with connection to mailjet or pull down and connect your testing mautic instance to your testing mailjet account.  

2.) After implementing the patch (respectively after pulling down and established your mailjet test environment) clear the cache of your mautic test instance with mailjet.

3.) Test the solution by sending an email via mautic to a "softbounce simulator" or an email account, which you configured as „full"

4.) When the softbounced-response of mailjet is back (remember, that can take up to three days; see also step two under I.) Reproduce the problem), check the `comment` field in the table `lead_donotcontact` for the soft_bounced email. There you should then see: `SOFT: : : „comment-message-of-mailjet"`.

5.) Test the solution by sending an email, that will get the status „hardbounced“ in mailjet.

6.) When the hardbounce-response of mailjet is back, check the `comment` field in the table `lead_donotcontact` for the hard_bounced email. There you should see: `HARD: „error_related_to text of mailjet“: „error text of mailjet“ `

7.) Test the solution by sending an email, that will get the status „blocked“ in mailjet.

8.) When the blocked-response of mailjet is back, check the `comment` field in the table `lead_donotcontact` for the hard_bounced email. There you should see: `BLOCKED: „error_related_to text of mailjet“: „error text of mailjet“ `
